packages/compass-aggregations/src/utils/cancellablePromise.ts

```diff
@@ -0,0 +1,50 @@
+const OPERATION_CANCELLED_MESSAGE = 'The operation was cancelled.';
+/*
+ * Return a promise you can race (just like a timeout from timeouts/promises).
+ * It will reject if abortSignal triggers before successSignal
+*/
+function abortablePromise(abortSignal: AbortSignal, successSignal: AbortSignal) {
+  let reject: (reason: unknown) => void;
+
+  const promise = new Promise<never>(function (_resolve, _reject) {
+    reject = _reject;
+  });
+
+  const abort = () => {
+    // if this task aborts it will never succeed, so clean up that event listener
+    // (abortSignal's event handler is already removed due to { once: true })
+    successSignal.removeEventListener('abort', succeed);
+
+    reject(new Error(OPERATION_CANCELLED_MESSAGE));
+  };
+
+  const succeed = () => {
+    // if this task succeeds it will never abort, so clean up that event listener
+    // (successSignal's event handler is already removed due to { once: true })
+    abortSignal.removeEventListener('abort', abort);
+  };
+
+  abortSignal.addEventListener('abort', abort, { once: true });
+  successSignal.addEventListener('abort', succeed, { once: true });
```

Comment:
We should probably also check whether `signal.aborted` is already true before adding these listeners.